src/parser/core/modules/Combos.tsx

```diff
@@ -0,0 +1,252 @@
+				return true
+			}
+
+			// Combo action that isn't a starter, that's a paddlin'
+			if (combo.from) {
+				this.recordUncomboedGcd(event)
+				return false
+			}
+		}
+
+		// Incorrect combo action, that's a paddlin'
+		if (combo.from !== this.lastAction) {
+			this.recordBrokenCombo(event, this.currentComboChain)
+			return combo.start // It's a combo if the action is the start of one
+		}
+
+		// Combo continued correctly
+		this.fabricateComboEvent(event)
+		// If it's a finisher, reset the combo
+		return !combo.end
+	}
+
+	private onCast(event: CastEvent) {
+		const action = getAction(event.ability.guid) as TODO // Should be an Action type
+
+		if (!action) {
+			return
+		}
+
+		if (action.onGcd) {
+			if (event.timestamp - this.lastGcdTime > GCD_TIMEOUT_MILLIS) {
+				// If we've had enough downtime between GCDs to let the combo expire, reset the state so we don't count erroneous combo breaks
+				this.currentComboChain = []
+			}
+
+			this.lastGcdTime = event.timestamp
+		}
+
+		// If it's a combo action, run it through the combo checking logic
+		if (action.combo) {
+			const continueCombo = this.checkCombo(action.combo, event)
+			if (continueCombo) {
+				this.currentComboChain.push(event)
+			} else {
+				this.currentComboChain = []
+			}
+		}
+
+		if (action.breaksCombo && this.lastAction !== null) {
+			// Combo breaking action, that's a paddlin'
+			this.recordBrokenCombo(event, this.currentComboChain)
+		}
+	}
+
+	private onComplete() {
+		if (this.addJobSpecificSuggestions(this.comboBreakers, this.uncomboedGcds)) {
+			return
+		}
+		if (this.issues.length > 0) {
+			this.suggestions.add(new Suggestion({
+				icon: this.constructor.suggestionIcon,
+				content: <Trans id="core.combos.content">
+					Avoid misusing your combo GCDs at the wrong combo step or breaking existing combos with non-combo
+					GCDs. Breaking combos can cost you significant amounts DPS as well as important secondary effects.
+				</Trans>,
+				severity: SEVERITY.MEDIUM, // TODO
+				why: <Plural
+					id="core.combos.why"
+					value={this.issues.length}
+					one="You misused # combo action."
+					other="You misused # combo actions."
+				/>,
+			}))
+		}
+	}
+
+	addJobSpecificSuggestions(comboBreakers: CastEvent[], uncomboedGcds: CastEvent[]) {
+		// To be overridden by subclasses. This is called in _onComplete() and passed two arrays of event objects - one for events that
+		// broke combos, and one for combo GCDs used outside of combos. Subclassing modules can add job-specific suggestions based on
+		// what particular actions were misused and when in the fight.
+		// The overriding module should return true if the default suggestion is not wanted
+		return false
+	}
+
+	output(): React.ReactNode {
+		if (this.issues.length > 0) {
+			return <RotationTable
+				notes={[
+					{
+						header: <Trans id="core.combos.rotationtable.header.reason">Reason</Trans>,
+						accessor: 'reason',
+					},
+				]}
+				data={this.issues
```

Comment:
would rather this was calculated outside the component

could flatten this entire block by reversing the condition too

Reply:
I don't even remember why I did it this way... considering I try to always go put the expected value as the straight up return and catch edge cases beforehand...